providers/base/bin/mei.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+
+import array
+import uuid
+import fcntl
+import struct
+import os
+from argparse import ArgumentParser
+
+
+DEFAULT_MEI_NODE = "mei0"
+
+
+class MEI_INTERFACE():
+
+    IOCTL_MEI_CONNECT_CLIENT = 0xc0104801
+
+    def __init__(self):
+        self._mei_obj = None
+
+    def _get_mei(self):
+        path = "/dev"
+        devices = os.listdir(path)
+        if DEFAULT_MEI_NODE in devices:
+            return os.path.join(path, DEFAULT_MEI_NODE)
+        for device in devices:
+            if device.find("mei") != -1:
+                return os.path.join(path, device)
+
+    def open(self):
+        mei_path = self._get_mei()
+        if mei_path is None:
+            raise SystemExit("MEI interface not found")
+        print("connecting to {}".format(mei_path))
+        self._mei_obj = os.open(mei_path, os.O_RDWR)
+
+    def connect(self, str_uuid):
+        obj_uuid = uuid.UUID(str_uuid)
+        array_data = array.array("b", obj_uuid.bytes_le)
+        fcntl.ioctl(self._mei_obj,
+                    self.IOCTL_MEI_CONNECT_CLIENT,
+                    array_data, 1)
+        max_length, version = struct.unpack("<IB", array_data.tobytes()[:5])
+        return max_length, version
+
+    def write(self, msg_id):
+        data_write = struct.pack("I", msg_id)
+        result = os.write(self._mei_obj, data_write)
+        return result
+
+    def read(self, size):
+        data = os.read(self._mei_obj, size)
+        return data
+
+    def close(self):
+        os.close(self._mei_obj)
+
+
+def get_mei_firmware_version():
+    mei_fw_uuid = "8e6a6715-9abc-4043-88ef-9e39c6f63e0f"
+    mei_fw_ver_req = 0x000002FF
+    mei_fw_ver_rep_length = 28
```

Comment:
Not sure how these values are defined, nor where do they come from. If you have some public source, please add a comment above to provide more explanation.

Reply:
I've left some comment about these fixed variables.